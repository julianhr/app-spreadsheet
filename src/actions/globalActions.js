import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')
export const displayInputter = createAction('DISPLAY_INPUTTER')
