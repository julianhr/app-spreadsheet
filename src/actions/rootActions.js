import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')

export function moveActiveCell(key) {
  return (dispatch, getState) => {
    const { activeCell, rows, columns } = getState()
    const currRow = activeCell.codePointAt(0) - 'a'.codePointAt(0)
    const currCol = Number.parseInt(activeCell[1]) - 1
    let nextRow = currRow
    let nextCol = currCol

    if (key === 'ArrowUp') { nextRow = currRow - 1 }
    else if (key === 'ArrowRight') { nextCol = currCol + 1 }
    else if (key === 'ArrowDown') { nextRow = currRow + 1 }
    else if (key === 'ArrowLeft') { nextCol = currCol - 1 }

    if (Math.min(nextRow, nextCol) >= 0 && nextRow < rows && nextCol < columns) {
      nextRow = String.fromCodePoint('a'.codePointAt(0) + nextRow)
      nextCol = nextCol + 1
      dispatch(setActiveCell(`${nextRow}${nextCol}`))
    }
  }
}
