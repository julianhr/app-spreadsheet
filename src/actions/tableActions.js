import graph from '~/formulas/graph'
import Interpreter from '../formulas/Interpreter'


export function setCellData(location, entered) {
  const interpreter = new Interpreter(location)
  const result = interpreter.interpret(entered)


  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function replaceCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().table[location]

    graph.delVertex(location)

    if (value === undefined) {
      return Promise.resolve()
    }


    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
