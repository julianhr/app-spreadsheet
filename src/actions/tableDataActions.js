export function setCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().tableData[location]

    if (value === undefined) {
      return Promise.resolve()
    }

    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
