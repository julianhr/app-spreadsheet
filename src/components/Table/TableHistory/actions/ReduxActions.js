import reduxConnect from '../reduxConnect'


class ReduxActions {
  static reduxConnect = reduxConnect

  get reduxConnect() {
    return ReduxActions.reduxConnect
  }

  getCellData(location) {
    return this.reduxConnect.getState().tableData[location]
  }
}

export default ReduxActions
