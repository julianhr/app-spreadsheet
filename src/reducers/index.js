import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
