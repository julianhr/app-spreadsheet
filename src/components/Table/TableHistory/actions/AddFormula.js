import ReduxActions from './ReduxActions'


class AddFormula extends ReduxActions {
  constructor(location, oldVertex, newVertex) {
    super()
    this.location = location
    this.entered = newVertex.entered
    this.result = newVertex.result
    this.previous = oldVertex && {
      entered: oldVertex.entered,
      result: oldVertex.result,
    }
  }

  execute() {
    this.reduxConnect.fn.setCellData(this.location, this.entered, this.result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (this.previous) {
      const { entered, result } = this.previous
      this.reduxConnect.fn.setCellData(this.location, entered, result)
    } else {
      this.reduxConnect.fn.clearCellData(this.location)
    }

    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default AddFormula
