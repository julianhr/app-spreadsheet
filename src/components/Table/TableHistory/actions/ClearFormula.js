import ReduxActions from './ReduxActions'


class ClearFormula extends ReduxActions {
  constructor(location) {
    super()
    this.previous = this.getCellData(location)
    this.location = location
  }

  execute() {
    this.reduxConnect.fn.clearCellData(this.location)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (!this.previous) { return }
    const { entered, result } = this.previous
    this.reduxConnect.fn.setCellData(this.location, entered, result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default ClearFormula