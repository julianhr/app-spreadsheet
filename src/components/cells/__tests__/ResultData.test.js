import React from 'react'
import { create } from 'react-test-renderer'
import { render, fireEvent, cleanup } from 'react-testing-library'
import emotionSerializer from 'jest-emotion'

import ConnectedResultData, { ResultData } from '../ResultData'
import MockApp from '~/__tests__/__mocks__/MockApp'
import { appStoreGen } from '~/reducers'
import { setCellData } from '~/actions/tableActions'


const requiredProps = {
  // props
  location: 'B-2',
  onDoubleClick: jest.fn(),
  onKeyDownEditable: jest.fn(),
  // redux
  clearCellData: jest.fn(),
  setActiveCell: jest.fn(),
  result: 'test result',
}

const testProps = {
  ...requiredProps,
  // props
  isFocused: true,
}

const createApp = (props) => {
  jest.spyOn(ResultData.prototype, 'focusDataTag').mockReturnValueOnce(null)
  const wrapper = create(
    <MockApp>
      <ResultData {...props} />
    </MockApp>
  )

  const cell = wrapper.root.findByType(ResultData).instance
  cell.refDataTag.current = document.createElement('div')
  return wrapper
}

const renderApp = (props, customStore) => {
  const store = customStore || appStoreGen()
  const wrapper = render(
    <MockApp customStore={store}>
      <ConnectedResultData {...props} />
    </MockApp>
  )
  return [wrapper, store]
}

describe('ResultData', () => {
  afterEach(cleanup)

  describe('props', () => {
    test('all required props', () => {
      expect(() => createApp(testProps)).not.toThrow()
    })

    it('raises warning if required props is not present', () => {
      Object.keys(requiredProps).forEach(key => {
        const props = {...testProps}
        delete props[key]
        expect(() => createApp(props)).toThrow()
      })
    })
  })

  describe('#focusDataTag', () => {
    it('if props.isFocused = true then focus text tag on mount', () => {
      const props = {...testProps}
      props.isFocused = true
      renderApp(props)

      const tag = document.querySelector(`[data-location="${props.location}"`)
      expect(document.activeElement).toBe(tag)
    })

    it('if props.isFocused = false then do not focus text tag', () => {
      const props = {...testProps}
      props.isFocused = false
      renderApp(props)

      const tag = document.querySelector(`[data-location="${props.location}"`)
      expect(document.activeElement).not.toBe(tag)
    })
  })

  describe('#handleOnKeyDown', () => {
    it('sets active cell to current one', async () => {
      const store = appStoreGen()
      expect(store.getState().global.activeCell).toBeNull()

      renderApp(testProps, store)
      const tag = document.querySelector(`[data-location="${testProps.location}"]`)
      fireEvent.keyDown(tag, { key: 'a' })
      expect(store.getState().global.activeCell).toEqual(testProps.location)
    })

    it('clears value if key pressed is Backspace or Delete', async () => {
      const store = appStoreGen()
      const { location } = testProps
      const entered = 'test entered'
      const result = entered

      const keys = [
        ['Backspace', false],
        ['Delete', false],
        ['a', true],
        ['1', true],
        [',', true]
      ]
      renderApp(testProps, store)
      const tag = document.querySelector(`[data-location="${testProps.location}"]`)

      for (let [key, expected] of keys) {
        await store.dispatch(setCellData(location, entered, true, result))
        fireEvent.keyDown(tag, { key })
        const hasValue = Boolean(store.getState().table[testProps.location])
        expect(hasValue).toEqual(expected)
      }
    })

    it('calls #onKeyDownEditable if key pressed is printable symbol', () => {
      const keys = [
        ['a', 1],
        ['1', 2],
        [',', 3],
        ['Backspace', 3],
        ['Delete', 3],
      ]
      testProps.onKeyDownEditable.mockReset()
      renderApp(testProps)
      const tag = document.querySelector(`[data-location="${testProps.location}"]`)

      for (let [key, times] of keys) {
        fireEvent.keyDown(tag, { key })
        expect(testProps.onKeyDownEditable).toHaveBeenCalledTimes(times)
      }
    })
  })

  describe('functional tests', () => {
    it('snapshot with value as string', () => {
      expect.addSnapshotSerializer(emotionSerializer)
      render(<MockApp><ResultData {...testProps} /></MockApp>)
      const el = document.querySelector(`[data-location="${testProps.location}"]`)
      expect(el).toMatchSnapshot()
    })

    it('snapshot with value as number', () => {
      expect.addSnapshotSerializer(emotionSerializer)
      const props = {  ...testProps, result: 5 }
      render(<MockApp><ResultData {...props} /></MockApp>)
      const el = document.querySelector(`[data-location="${props.location}"`)
      expect(el).toMatchSnapshot()
    })
  })
})
