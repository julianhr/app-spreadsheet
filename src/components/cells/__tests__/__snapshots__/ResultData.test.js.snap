// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultData functional tests snapshot with value as number 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: none;
  border: 2px solid transparent;
  line-height: 1.1em;
  height: 100%;
  width: 100%;
  padding: 2px;
  white-space: nowrap;
  text-overflow: clip;
  overflow: hidden;
}

.emotion-0:focus {
  border: 2px solid salmon;
}

<div
  class="emotion-0 emotion-1"
  data-cell="result"
  data-location="B-2"
  tabindex="0"
>
  5
</div>
`;

exports[`ResultData functional tests snapshot with value as string 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: none;
  border: 2px solid transparent;
  line-height: 1.1em;
  height: 100%;
  width: 100%;
  padding: 2px;
}

.emotion-0:focus {
  border: 2px solid salmon;
}

<div
  class="emotion-0 emotion-1"
  data-cell="result"
  data-location="B-2"
  tabindex="0"
>
  test result
</div>
`;
